src/IO/S3Common.cpp

```diff
@@ -54,6 +55,47 @@ class AWSLogger final : public Aws::Utils::Logging::LogSystemInterface
 private:
     Poco::Logger * log = &Poco::Logger::get("AWSClient");
 };
+
+class S3AuthSigner : public Aws::Client::AWSAuthV4Signer
+{
+public:
+    S3AuthSigner(
+        const Aws::Client::ClientConfiguration & clientConfiguration,
+        const Aws::Auth::AWSCredentials & credentials,
+        const std::vector<std::pair<const String, const String>> & headers_)
+        : Aws::Client::AWSAuthV4Signer(
+            std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(credentials),
+            "s3",
+            clientConfiguration.region,
+            Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Never,
+            false)
+        , headers(headers_)
+    {
+    }
+
+    bool SignRequest(Aws::Http::HttpRequest & request, const char * region, bool signBody) const override
+    {
+        auto result = Aws::Client::AWSAuthV4Signer::SignRequest(request, region, signBody);
+        for (const auto & header : headers)
+            request.SetHeaderValue(header.first, header.second);
+        return result;
+    }
+
+    bool PresignRequest(
+        Aws::Http::HttpRequest & request,
+        const char * region,
+        const char * serviceName,
+        long long expirationTimeInSeconds) const override // NOLINT
+    {
+        auto result = Aws::Client::AWSAuthV4Signer::PresignRequest(request, region, serviceName, expirationTimeInSeconds);
+        for (const auto & header : headers)
+            request.SetHeaderValue(header.first, header.second);
+        return result;
+    }
+
+private:
+    const std::vector<std::pair<const String, const String>> headers;
```

Comment:
Suspicious type. Better to create structure and use `using XXX = std::vector<...>`.